### PR TITLE
Feature/block27 Provision AKS environment pipeline

### LINKED ISSUE #27: Block: Provision AKS environment
As someone setting up the project infrastructure, I want to have a detailed guide on how to provision the environment so that I can get this done fast and without errors.

Same as #10 but for Azure AKS and without the part of evaluating Rancher.

### PR DESCRIPTION
**Pull request by jbellver99**
Closes #27

### CODE REVIEW

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 1]
```diff
@@ -1,197 +1,261 @@
...
-        i) imageName=${OPTARG};;
+ARGS=$*
+FLAGS=$(getopt -a --options c:n:d:a:b:l:i:u:p: --long "config-file:,pipeline-name:,local-directory:,artifact-path:,target-branch:,language:,build-pipeline-name:,sonar-url:,sonar-token:,image-name:,user:,password:,resource-group:,storage-account:,storage-container:,cluster-name:,s3-bucket:,s3-key-path:" -- "$@")
+eval set -- "$FLAGS"
+while true; do
+    case "$1" in
+        -c | --config-file)     configFile=$2; shift 2;;
+        -n | --pipeline-name)   pipelineName=$2; shift 2;;
+        -d | --local-directory) localDirectory=$2; shift 2;;
+        -a | --artifact-path)   artifactPath=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:10]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 2]
```diff
@@ -1,197 +1,261 @@
...
+        -b | --target-branch)   targetBranch=$2; shift 2;;
+        -l | --language)        language=$2; shift 2;;
+        --build-pipeline-name)  buildPipelineName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:13]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 3]
```diff
@@ -1,197 +1,261 @@
...
+        --sonar-url)            sonarUrl=$2; shift 2;;
+        --sonar-token)          sonarToken=$2; shift 2;;
+        -i | --image-name)      imageName=$2; shift 2;;
+        -u | --user)            user=$2; shift 2;;
+        -p | --password)        password=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:18]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 4]
```diff
@@ -1,197 +1,261 @@
...
+        --resource-group)       resourceGroupName=$2; shift 2;;
+        --storage-account)      storageAccountName=$2; shift 2;;
+        --storage-container)    storageContainerName=$2; shift 2;;
+        --cluster-name)         clusterName=$2; shift 2;;
+        --s3-bucket)            s3Bucket=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:23]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 5]
```diff
@@ -1,197 +1,261 @@
...
+        --s3-key-path)          s3KeyPath=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:24]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 6]
```diff
@@ -1,197 +1,261 @@
...
+        -b | --target-branch)   targetBranch=$2; shift 2;;
+        -l | --language)        language=$2; shift 2;;
+        --build-pipeline-name)  buildPipelineName=$2; shift 2;;
+        --sonar-url)            sonarUrl=$2; shift 2;;
+        --sonar-token)          sonarToken=$2; shift 2;;
+        -i | --image-name)      imageName=$2; shift 2;;
+        -u | --user)            user=$2; shift 2;;
+        -p | --password)        password=$2; shift 2;;
+        --resource-group)       resourceGroupName=$2; shift 2;;
+        --storage-account)      storageAccountName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:20]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 7]
```diff
@@ -1,197 +1,261 @@
...
+        --storage-container)    storageContainerName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:21]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 8]
```diff
@@ -1,197 +1,261 @@
...
+        -n | --pipeline-name)   pipelineName=$2; shift 2;;
+        -d | --local-directory) localDirectory=$2; shift 2;;
+        -a | --artifact-path)   artifactPath=$2; shift 2;;
+        -b | --target-branch)   targetBranch=$2; shift 2;;
+        -l | --language)        language=$2; shift 2;;
+        --build-pipeline-name)  buildPipelineName=$2; shift 2;;
+        --sonar-url)            sonarUrl=$2; shift 2;;
+        --sonar-token)          sonarToken=$2; shift 2;;
+        -i | --image-name)      imageName=$2; shift 2;;
+        -u | --user)            user=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:17]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 9]
```diff
@@ -1,197 +1,261 @@
...
+FLAGS=$(getopt -a --options c:n:d:a:b:l:i:u:p: --long "config-file:,pipeline-name:,local-directory:,artifact-path:,target-branch:,language:,build-pipeline-name:,sonar-url:,sonar-token:,image-name:,user:,password:,resource-group:,storage-account:,storage-container:,cluster-name:,s3-bucket:,s3-key-path:" -- "$@")
+eval set -- "$FLAGS"
+while true; do
+    case "$1" in
+        -c | --config-file)     configFile=$2; shift 2;;
+        -n | --pipeline-name)   pipelineName=$2; shift 2;;
+        -d | --local-directory) localDirectory=$2; shift 2;;
+        -a | --artifact-path)   artifactPath=$2; shift 2;;
+        -b | --target-branch)   targetBranch=$2; shift 2;;
+        -l | --language)        language=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:12]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+ip=$(kubectl get -o json svc nginx-stable-nginx-ingress --namespace nginx-ingress --kubeconfig $1/kubeconfig | python -c "import sys, json; print(json.load(sys.stdin)['status']['loadBalancer']['ingress'][0]['ip'])")
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:2]: @sonatype-lift ignore

File: scripts/environment-provisioning/azure/aks/set-config.sh [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+ARGS=$*
+/bin/sh ../../../set-config.sh $ARGS
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/environment-provisioning/azure/aks/set-config.sh:3]: @sonatype-lift ignore

File: documentation/azure-devops/setup-aks-provisioning-pipeline.asciidoc [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+```
+./pipeline_generator.sh -c ./templates/aks/aks-pipeline.cfg -n aks-provisioning --resource-group hangar-aks-rg --storage-account hangarstatefiles --storage-container tfstatefile -d C:/Users/$USERNAME/Desktop/quarkus-project -b develop -w
+```
+
+=== Appendix: Interacting with the cluster
+
+NOTE: Make sure to have https://kubernetes.io/docs/tasks/tools/[kubectl] installed in order to interact with your cluster.
+
+In order to interact with your cluster download the artifact `kubeconfig` generated by the pipeline and execute the following command:
```
Comment (albertdb) [inline documentation/azure-devops/setup-aks-provisioning-pipeline.asciidoc:83]:
```suggestion
In order to interact with your cluster you will need to download the artifact `kubeconfig` generated by the cluster provisioning pipeline on the location it is expected by default (`~/.kube/config`) or either:
```

File: documentation/azure-devops/setup-aks-provisioning-pipeline.asciidoc [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+```
+kubectl <kubectl command> --kubeconfig <path to downloaded kubeconfig>
```
Comment (albertdb) [inline documentation/azure-devops/setup-aks-provisioning-pipeline.asciidoc:86]:
```suggestion
# via environment variable (you can add this on your profile)
export KUBECONFIG=<path to downloaded kubeconfig>
kubectl <command>

# via command-line flag
kubectl <command> --kubeconfig=<path to downloaded kubeconfig>
```

File: scripts/environment-provisioning/azure/aks/main.tf [hunk 14]
```diff
@@ -0,0 +1,36 @@
+terraform {
+    required_providers {
+        azurerm = {
+            source = "hashicorp/azurerm"
```
Comment (albertdb) [inline scripts/environment-provisioning/azure/aks/main.tf:4]: Please set version.

File: scripts/environment-provisioning/azure/aks/variables.tf [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    type = string
+}
+
+variable "instance_type" {
+    description = "Type of instance used"
+    type = string
+}
+
+variable "worker_node_count" {
+    description = "Number of nodes for the cluster"
```
Comment (albertdb) [inline scripts/environment-provisioning/azure/aks/variables.tf:22]:
```suggestion
    description = "Number of worker nodes for the cluster"
```

File: scripts/environment-provisioning/azure/aks/variables.tf [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    type = string
+}
+
+variable "dns_prefix" {
+    description = "DNS name prefix to use with the hosted Kubernetes API server FQDN"
```
Comment (albertdb) [inline scripts/environment-provisioning/azure/aks/variables.tf:27]: Could you please explain what it this? Not totally clear for me at least given the description.

File: scripts/pipelines/azure-devops/templates/aks/aks-pipeline.cfg [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    # Copy the terraform files.
+    cd "${hangarPath}/${terraformTemplatesPath}"
+    cp * "${localDirectory}/${terraformPath}"
+
+    # Check if an extra artifact to store is supplied.
+    if test ! -z "$artifactPath"
+    then
+        # Add the extra step to the YAML.
+        cat "${hangarPath}/${templatesPath}/store-extra-path.yml" >> "${localDirectory}/${scriptFilePath}/${yamlFile}"
+    fi
```
Comment (albertdb) [inline scripts/pipelines/azure-devops/templates/aks/aks-pipeline.cfg:35]:
```suggestion
```

Not needed anymore here.

File: scripts/pipelines/azure-devops/templates/aks/aks-provisioning.yml.template [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+- task: Bash@3
+  displayName: 'Install nginx-ingress controller'
+  inputs:
+    targetType: 'inline'
+    script: |
+      helm repo add nginx-stable https://helm.nginx.com/stable
+
+      helm repo update
+
+      helm install nginx-stable nginx-stable/nginx-ingress --set controller.setAsDefaultIngress=true --namespace nginx-ingress --create-namespace --kubeconfig $(terraformWorkingDirectory)/kubeconfig
```
Comment (albertdb) [inline scripts/pipelines/azure-devops/templates/aks/aks-provisioning.yml.template:50]: This is using a deprecated helm chart. Please look at EKS yaml.

File: scripts/pipelines/azure-devops/templates/aks/aks-provisioning.yml.template [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+
+- task: Bash@3
+  displayName: 'Wait for the ingress controller IP'
+  inputs:
+    targetType: 'inline'
+    script: 'sleep 60'
```
Comment (albertdb) [inline scripts/pipelines/azure-devops/templates/aks/aks-provisioning.yml.template:56]: Use proper wait.

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 20]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+ip="$(kubectl get -o json svc nginx-stable-nginx-ingress --namespace nginx-ingress --kubeconfig $1/kubeconfig | python -c "import sys, json; print(json.load(sys.stdin)['status']['loadBalancer']['ingress'][0]['ip'])")"
+
+url=$2
+length=${#url}
+dnsname=${url:22:length-22-1}
```
Comment (albertdb) [inline scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:6]: Maybe this could easily break?

**Review by albertdb — CHANGES_REQUESTED** [review]

File: documentation/azure-devops/setup-aks-provisioning-pipeline.asciidoc [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+* This script will commit and push the corresponding YAML template into your repository, so please be sure your local repository is up-to-date (i.e you have pulled the latest changes with `git pull`).
+
+== Creating the pipeline using provided script
+
+Before executing the script you will need to customize some input variables about the environment. To do so, you can either edit `terraform.tfvars` file or take advantage of the `set-config.sh` script located at `/scripts/environment-provisioning/azure/aks`, which allows you to create or update values for the required variables, passing them as flags. As a full example:
+
+```
+./set-config.sh --cluster_name <cluster name> --location <location> --resource_group_name <resource group name> --instance_type <worker instance type> --worker_node_count <number of worker nodes> --dns_prefix <dns prefix>
```
Comment (albertdb) [inline documentation/azure-devops/setup-aks-provisioning-pipeline.asciidoc:40]:
```suggestion
./set-terraform-variables.sh --cluster_name <cluster name> --location <location> --resource_group_name <resource group name> --instance_type <worker instance type> --worker_node_count <number of worker nodes> --dns_prefix <dns prefix>
```

**Review by albertdb — CHANGES_REQUESTED** [review]

File: scripts/environment-provisioning/azure/aks/set-terraform-variables.sh [hunk 22]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+/bin/sh ../../../set-config.sh $*
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/environment-provisioning/azure/aks/set-terraform-variables.sh:2]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/environment-provisioning/azure/aks/set-terraform-variables.sh:2]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/environment-provisioning/azure/aks/set-terraform-variables.sh:2]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/environment-provisioning/azure/aks/set-terraform-variables.sh:2]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 23]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+ip=$(kubectl get svc nginx-ingress-nginx-ingress-controller --namespace nginx-ingress --kubeconfig $1/kubeconfig -o jsonpath='{.status.loadBalancer.ingress[0].ip}' &> /dev/null)
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:2]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+
+while test -z "$ip"
+do
+    sleep 5s
+    ip=$(kubectl get svc nginx-ingress-nginx-ingress-controller --namespace nginx-ingress --kubeconfig $1/kubeconfig -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:7]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+url=$2
+length=${#url}
+# Obtain the organization name by taking away the first 22 characters (https://dev.azure.com/) and the last /.
+dnsname=${url:22:length-22-1}
+
+ipname=$(az network public-ip list --query "[?ipAddress!=null]|[?contains(ipAddress, '$ip')].[name]" --output tsv)
+
+iprg=$(az network public-ip list --query "[?ipAddress!=null]|[?contains(ipAddress, '$ip')].[resourceGroup]" --output tsv)
+
+az network public-ip update --resource-group $iprg --name $ipname --dns-name $dnsname
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:19]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:19]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:19]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+
+dns=$(az network public-ip show --resource-group $iprg --name $ipname --query "[dnsSettings.fqdn]" --output tsv)
```
[the remark this answers is not in the document]
Comment (jbellver99) [reply scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:21]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 27]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+ip="$(kubectl get svc nginx-ingress-nginx-ingress-controller --namespace nginx-ingress --kubeconfig "$1"/kubeconfig -o jsonpath='{.status.loadBalancer.ingress[0].ip}')"
```
Comment (albertdb) [inline scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:2]:
```suggestion
ip="$(kubectl get svc nginx-ingress-nginx-ingress-controller --namespace nginx-ingress --kubeconfig "$1/kubeconfig" -o jsonpath='{.status.loadBalancer.ingress[0].ip}')"
```

Although better pass KUBECONFIG env variable.

File: scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+
+while test -z "$ip"
+do
+    sleep 5s
+    ip="$(kubectl get svc nginx-ingress-nginx-ingress-controller --namespace nginx-ingress --kubeconfig "$1"/kubeconfig -o jsonpath='{.status.loadBalancer.ingress[0].ip}')"
```
Comment (albertdb) [inline scripts/pipelines/azure-devops/templates/aks/obtain-dns.sh:7]:
```suggestion
    ip="$(kubectl get svc nginx-ingress-nginx-ingress-controller --namespace nginx-ingress --kubeconfig "$1/kubeconfig" -o jsonpath='{.status.loadBalancer.ingress[0].ip}')"
```

Although better pass KUBECONFIG env variable.

**Review by albertdb — CHANGES_REQUESTED** [review]

**Review by albertdb — CHANGES_REQUESTED** [review]
Add destroy option to pipeline, in sync with EKS.

**Review by albertdb — APPROVED** [review]

**Pull request merged**